library/std/src/sys/unix/os.rs

```diff
@@ -493,17 +494,16 @@ pub unsafe fn environ() -> *mut *const *const c_char {
     &mut environ
 }
 
-pub unsafe fn env_lock() -> StaticMutexGuard {
-    // It is UB to attempt to acquire this mutex reentrantly!
-    static ENV_LOCK: StaticMutex = StaticMutex::new();
-    ENV_LOCK.lock()
+pub unsafe fn env_rwlock(readonly: bool) -> RWLockGuard {
+    static ENV_LOCK: RWLock = RWLock::new();
+    if readonly { ENV_LOCK.read_with_guard() } else { ENV_LOCK.write_with_guard() }
 }
 
 /// Returns a vector of (variable, value) byte-vector pairs for all the
 /// environment variables of the current process.
 pub fn env() -> Env {
     unsafe {
-        let _guard = env_lock();
+        let _guard = env_rwlock(true);
         let mut environ = *environ();
```

Comment:
`environ()` creates a `&mut` to `static mut environ`. With your change, this can happen multiple times at once, breaking the guarantees of `&mut`.

Reply:
Using `&raw mut environ` would be enough fix that, right? I.e. `*mut` is fine as long as it's not turned into a `&mut`.

Reply:
That should be fine. You could use ptr::addr_of_mut!() for that.